contracts/governance/Staking/WeightedStaking.sol

```diff
@@ -333,6 +333,133 @@ contract WeightedStaking is Checkpoints {
+			}
+		}
+	}
+
+	/**
+	 * @notice Compute the voting power for a specific date.
+	 * Power = stake * weight
+	 * TODO: WeightedStaking::weightedVestingStakeByDate should probably better
+	 * be internal instead of a public function.
+	 * @param date The staking date to compute the power for.
+	 * @param startDate The date for which we need to know the power of the stake.
+	 * @param blockNumber The block number, needed for checkpointing.
+	 * */
+	function weightedVestingStakeByDate(
+		address account,
+		uint256 date,
+		uint256 startDate,
+		uint256 blockNumber
+	) public view returns (uint96 power) {
+		uint96 staked = _getPriorUserVestingStakeByDate(account, date, blockNumber);
+		if (staked > 0) {
+			uint96 weight = computeWeightByDate(date, startDate);
+			power = mul96(staked, weight, "WeightedStaking::weightedVestingStakeByDate: multiplication overflow") / WEIGHT_FACTOR;
+		} else {
+			power = 0;
+		}
+	}
+
+	/**
+	 * @notice Determine the prior number of vested stake for an account until a
+	 * certain lock date as of a block number.
+	 * @dev Block number must be a finalized block or else this function
+	 * will revert to prevent misinformation.
+	 * @param account The address of the account to check.
+	 * @param date The lock date.
+	 * @param blockNumber The block number to get the vote balance at.
+	 * @return The number of votes the account had as of the given block.
+	 * */
+	function getPriorUserVestingStakeByDate(
+		address account,
+		uint256 date,
+		uint256 blockNumber
+	) external view returns (uint96) {
+		uint96 priorStake = _getPriorUserVestingStakeByDate(account, date, blockNumber);
+		// @dev we need to modify function in order to workaround issue with Vesting.withdrawTokens:
+		//		return 1 instead of 0 if message sender is a contract.
+		if (priorStake == 0 && _isVestingContract()) {
+			priorStake = 1;
+		}
+		return priorStake;
+	}
+
+	/**
+	 * @notice Determine the prior number of vested stake for an account until a
+	 * 		certain lock date as of a block number.
+	 * @dev All functions of Staking contract use this internal version,
+	 * 		we need to modify public function in order to workaround issue with Vesting.withdrawTokens:
+	 * return 1 instead of 0 if message sender is a contract.
+	 * */
+	function _getPriorUserVestingStakeByDate(
+		address account,
+		uint256 date,
+		uint256 blockNumber
+	) internal view returns (uint96) {
+		require(blockNumber < block.number, "WeightedStaking::getPriorUserVestingStakeByDate: not yet determined");
+
+		date = _adjustDateForOrigin(date);
+		uint32 nCheckpoints = numUserVestingCheckpoints[account][date];
+		if (nCheckpoints == 0) {
+			return 0;
+		}
+
+		/// @dev First check most recent balance.
+		if (userVestingCheckpoints[account][date][nCheckpoints - 1].fromBlock <= blockNumber) {
+			return userVestingCheckpoints[account][date][nCheckpoints - 1].stake;
```

Comment:
It would be better to create a new storage variable that will point to `userVestingCheckpoints[account][date][nCheckpoints - 1]`. This will help you not to calculate the hashes of the mapping keys twice.